libraries/standard/mqtt/src/mqtt.c

```diff
@@ -75,6 +77,387 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+        IotLogInfoWithArgs( "State record updated. New state=%d.",
+                            publishRecordState );
+
+        /* Send PUBACK or PUBREC if necessary. */
+        status = sendPublishAcks( pContext,
+                                  packetIdentifier,
+                                  &publishRecordState );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* TODO: Should the publish info be passed instead? */
+        pContext->callbacks.appCallback( pContext, pIncomingPacket );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Handle received MQTT ack.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] pIncomingPacket Incoming packet.
+ *
+ * @return MQTTSuccess, MQTTIllegalState, or deserialization error.
+ */
+static MQTTStatus_t handleIncomingAck( MQTTContext_t * const pContext,
+                                       MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    /* Need a dummy variable for MQTT_DeserializeAck(). */
+    bool sessionPresent = false;
+    MQTTPubAckType_t ackType;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    switch( pIncomingPacket->type )
+    {
+        case MQTT_PACKET_TYPE_PUBACK:
+        case MQTT_PACKET_TYPE_PUBREC:
+        case MQTT_PACKET_TYPE_PUBREL:
+        case MQTT_PACKET_TYPE_PUBCOMP:
+            ackType = getAckFromPacketType( pIncomingPacket->type );
+            status = MQTT_DeserializeAck( pIncomingPacket, &packetIdentifier,  &sessionPresent );
+            IotLogInfoWithArgs( "Ack packet deserialized with result: %d.",
+                                status );
+            if( status == MQTTSuccess )
+            {
+                publishRecordState = MQTT_UpdateStateAck( pContext,
+                                                          packetIdentifier,
+                                                          ackType,
+                                                          MQTT_RECEIVE );
+                IotLogInfoWithArgs( "State record updated. New state=%d.",
+                                    publishRecordState );
+
+                /* Send PUBREL or PUBCOMP if necessary. */
+                status = sendPublishAcks( pContext,
+                                          packetIdentifier,
+                                          &publishRecordState );
+                if( status == MQTTSuccess )
+                {
+                    /* TODO: There doesn't seem to be a separate callback for acks. */
+                    pContext->callbacks.appCallback( pContext, pIncomingPacket );
+                }
+            }
+            break;
+        case MQTT_PACKET_TYPE_PINGRESP:
+            /* TODO: should this be handled here? pContext->waitingForPingResp = false; */
```

Comment:
We need to call app callback for ping response. Since this loop does not process keep-alive ( does not send ping requests) the responses will be handled by the application.

Reply:
Added both app callback and keep alive. Not sure if this was your intention